cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -286,4 +296,48 @@ public void processMessage(MetadataMessage message) {
       }
     }
   }
+
+  private class MetadataOperationProcessor implements MetadataMessageProcessor {
+
+    @Override
+    public void processMessage(MetadataMessage message)  {
+      MetadataOperation operation = message.getPayload(GSON, MetadataOperation.class);
+      Metadata metadata = operation.getMetadata();
+      MetadataEntity entity = operation.getEntity();
+
+      // TODO: Authorize that the operation is allowed. Currently MetadataMessage does not carry user info
+      switch (operation.getType()) {
+        case PUT: {
+          LOG.trace("Received PUT for entity {}: {}", entity, metadata);
+          try {
+            if (metadata.getProperties() != null && !metadata.getProperties().isEmpty()) {
+              metadataAdmin.addProperties(entity, metadata.getProperties());
+            }
+            if (metadata.getTags() != null && !metadata.getTags().isEmpty()) {
+              Set<String> toAdd = metadata.getTags();
+              metadataAdmin.addTags(entity, toAdd.toArray(new String[toAdd.size()]));
+
+            }
+          } catch (InvalidMetadataException e) {
+            LOG.warn("Received invalid metadata operation from TMS: {}", GSON.toJson(message.getRawPayload()), e);
+          }
+          break;
+        }
+        case DELETE: {
+          LOG.trace("Received DELETE for entity {}: {}", entity, metadata);
+          if (metadata.getProperties() != null && !metadata.getProperties().isEmpty()) {
+            Set<String> toRemove = metadata.getProperties().keySet();
+            metadataAdmin.removeProperties(entity, toRemove.toArray(new String[toRemove.size()]));
+          }
+          if (metadata.getTags() != null && !metadata.getTags().isEmpty()) {
+            Set<String> toRemove = metadata.getTags();
+            metadataAdmin.removeTags(entity, toRemove.toArray(new String[toRemove.size()]));
+          }
+          break;
+        }
+        default:
+          throw new IllegalStateException("Unknown MetadataOperation type " + operation.getType());
```

Comment:
You probably also want to just log and ignore unknown operation types instead?

Reply:
right, otherwise this will cause failures.